### PR TITLE
[RecoEgamma/EgammaTools] Updates to EnergyScaleCorrection class for scales and smearings in UL

### PR DESCRIPTION
**Pull request by jainshilpi**
This PR takes care of the updated txt file format which stores the scales and smearing corrections. 
Updates to following codes have been done:

(1)  EnergyScaleCorrection.cc/h - As Sam suggested, the code can now determine the txt file format and reads accordingly
(2) calibratedEgammas_cff.py - now the default txt file to be read is the UL one. For the last re-reco of 2017, run2_miniAOD_94XFall17 is used
(3) Small change to PhotonEnergyCalibrator/ElectronEnergyCalibrator have been done to avoid problems at the boundary. 

The related txt files can be found here: https://github.com/jainshilpi/EgammaAnalysis-ElectronTools/tree/UL2017SSV2/ScalesSmearings (*24Feb*)

runMatrix passed all the tests.

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoEgamma/EgammaTools/interface/EnergyScaleCorrection.h [hunk 1]
```diff
@@ -81,7 +81,7 @@ class EnergyScaleCorrection {
   public:
     CorrectionCategory(const std::string& category, int runnrMin = 0, int runnrMax = 999999);
     CorrectionCategory(
-        const unsigned int runnr, const float et, const float eta, const float r9, const unsigned int gainSeed)
+        const unsigned int runnr, const double et, const double eta, const double r9, const unsigned int gainSeed)
```
Comment (Sam-Harper) [inline RecoEgamma/EgammaTools/interface/EnergyScaleCorrection.h:84]: A general question, why all the changes from floats to doubles?

Comment (jainshilpi) [reply RecoEgamma/EgammaTools/interface/EnergyScaleCorrection.h:84]: I saw problems at boundaries of the corrections (e.g. eta=2.5000001 which was at a precision of 10^-7). So I had to update from float to double.

**Comment by jainshilpi** [comment]
thanks to Sam for raising this point - I realized I had missed one protection against boundary which was already present in the code for previous file formats. this is included for the new file format as well and all float are hence moved from double --> floats. This leaves PhotonEnergyCalibrator and ElectronEnergyCalibrator files unchanged.

**Comment by perrotta** [comment]
@cmsbuild please test

**Comment by perrotta** [comment]
please abort
(not sure the last commit had been included)

**Comment by perrotta** [comment]
@cmsbuild please test

**Review by slava77 — COMMENTED** [review]
apparently there is no workflow in the jenkins tests where the calibrated ele/pho are enabled for this new calib (except for run2_miniAOD_94XFall17 and run2_miniAOD_80XLegacy which explicitly do not use it).

Is there a workflow in the matrix to test this new correction?

**Comment by perrotta** [comment]
please test

**Comment by srimanob** [comment]
> Is there a workflow in the matrix to test this new correction?
@slava77 @jainshilpi Do you need new workflows to test?

**Comment by jainshilpi** [comment]
> > Is there a workflow in the matrix to test this new correction?
> > @slava77 @perrotta  @jainshilpi  Do you need new workflows to test?

These new UL corrections should be applied by default if not in those two eras. Just to mention that in EGM, we have already validated these corrections and can be found here (using EgammaPostRecoTools): https://shilpi-afs.web.cern.ch/shilpi-afs/13TeV/2020/SSvalidation/UL2017/30marchV2/

**Comment by slava77** [comment]
> These new UL corrections should be applied by default if not in those two eras.

there were no changes in comparisons, which suggests to me (also based on git grep of configs in the release) that only the two eras have the calibrated photons. So, it looks like there is no default running.

**Comment by silviodonato** [comment]
> > These new UL corrections should be applied by default if not in those two eras.
> 
> there were no changes in comparisons, which suggests to me (also based on git grep of configs in the release) that only the two eras have the calibrated photons/electrons. So, it looks like there is no default running.

@jainshilpi please note Slava's comment

**Comment by jainshilpi** [comment]
> > > These new UL corrections should be applied by default if not in those two eras.
> > 
> > 
> > there were no changes in comparisons, which suggests to me (also based on git grep of configs in the release) that only the two eras have the calibrated photons/electrons. So, it looks like there is no default running.
> 
> @jainshilpi please note Slava's comment

Hi Silvio, yes this was discussed in the PPD meetings and I discussed iwth Slava separately. I think Slava already is working on it and may have the workflow soon. So once it is done, I will proceed accordingly.

**Comment by slava77** [comment]
> The related txt files can be found here: https://github.com/jainshilpi/EgammaAnalysis-ElectronTools/tree/UL2017SSV2/ScalesSmearings (24Feb)

this requires a PR/update in https://github.com/cms-data/EgammaAnalysis-ElectronTools
Please make a PR for cms-data.
Thank you.

**Comment by jainshilpi** [comment]
I had forgotten to do that, sorry. Just did it.

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/29526	5f6a5e0
- code changes are in line with the PR description and the follow up review
- jenkins tests pass and comparisons with the baseline show no differences: the old reminiAOD results are preserved; the newly added smearing setup is not configured in the tests
- I checked the runtime locally in conjunction with #29756 cms-data/EgammaAnalysis-ElectronTools#8 and enabling the resolution/smearing with https://github.com/cms-sw/cmssw/compare/bfe0331e26bd125e10e4fb8a15afb80e148676f2...3258daf021a7d74665f28a79dbf0e9cd49b44584 : the new userFloats show up in slimmed electrons and photons as expected in the UL workflows

**Comment by silviodonato** [comment]
+1

**Pull request merged**